gitlab_runner_config.py

```diff
@@ -53,20 +56,39 @@ def generate_tags(instance, executor_type="", env=None):
     These tags are specified by runner configs and used by CI specs to run jobs
     on the appropriate host.
     """
-
-    tags = identifying_tags(instance)
-    if executor_type:
-        tags.append(executor_type)
-    if env:
-        tags += [os.environ[e] for e in env if e in os.environ]
+    properties = {
+        "hostname": HOSTNAME,
+        "executor_type": executor_type,
+        "instance": instance,
+        "env": []
+        }
+    if env: 
+        properties["env"] += [os.environ[e] for e in env if e in os.environ]
+    # if executor is batch, gather some more system info for tags
     if executor_type == "batch":
         if which("bsub"):
-            tags.append("lsf")
+            properties["scheduler"] = "lsf"
         elif which("salloc"):
-            tags.append("slurm")
+            properties["scheduler"] = "slurm"
         elif which("cqsub"):
-            tags.append("cobalt")
-    return tags
+            properties["scheduler"] = "cobalt"
+    try:
+        properties.update(capture_tags.capture_tags(instance, executor_type, env=env, tag_schema=tag_schema))
+    except NameError:
+        logger.info("Custom Tag Capture method not provided")
+    try:
+        if tag_schema:
+            validate(instance=properties, schema=tag_schema)
+        tags = list(properties.values())
+        for i in tags:
```

Comment:
Yes, I didn't mention this when we were talking but it is a case to consider: what do we do with nested lists?

I think the right answer (I'm open to counter examples), is to flatten all dictionary values into a `list` (we are only using keys for our sake during schema validation). Here's a working example of the idea:

```python
import json

ex = {"foo": ["bar", "baz", "bat"], "bam": {"bak": "baf", "bal": [1, 2, 3]}}


def flatten_values(d):
    "recursively collect dictionary values into a list"

    if isinstance(d, list):
        combined = []
        for item in d:
            combined += flatten_values(item)
        return combined
    elif isinstance(d, dict):
        combined = []
        for item in d.values():
            combined += flatten_values(item)
        return combined
    else:
        return [d]


print(f"Original:\n{json.dumps(ex, indent=2)}")
print(f"Flat:\n{json.dumps(flatten_values(ex), indent=2)}")
```